### DISCUSSION TITLE
(Jetbrains) IDE suggestions and Copilot suggestions can't be used both when using IdeaVim

### DISCUSSION THREAD

**Discussion by kjyv** [community/community#12714] [Copilot Conversations] [answered]
I'm using Jetbrains Rider 2021.3.3 and IdeaVim 1.10.0.

When typing something, Rider's completion popup opens and Copilot also suggests some completion at the same time and shows it in gray, so far so good. Now, I might want to use either a popup completion or the Copilot one but I can only choose the popup one.
When pressing Tab (or also a custom shortcut if I set it), this won't complete Copilot's suggestion but will insert Rider's suggestion (bound to Tab by default) or if I disable that shortcut, it just inserts a Tab character (and I don't want to disable that shortcut of course). If only Copilot suggests something but no popup is there, Tab will complete as it should.
Without IdeaVim enabled, it is possible to press Esc to close Rider's completion popup and then the Copilot suggestion will immediately be shown again and can be completed with Tab. With IdeaVim enabled both Rider's popup and the Copilot completion is being hidden by Esc because IdeaVim leaves insert mode and changes to normal mode (I could click somewhere to close the popup and it will show the Copilot suggestion but that is not very usable, a keyboard option is necessary).
Now, an idea would be to say IdeaVim should not leave insert mode but ask to press Esc again after first closing the popup but this behaviour was considered a bug that has been fixed: https://youtrack.jetbrains.com/issue/IDEA-209779.
So I guess another way that should be usable is that disabling Tab for popup completion should allow Tab to complete Copilot and/or a custom shortcut like Shift-Tab should complete Copilot's suggestion also while the popup is open (and also close it additionally if possible).
Currently it's quite tricky to actually use a good looking completion and it feels like something is always getting in the way.

Sorry for the long description, hope it's ok to follow.

**Comment by jansorg** [comment, marked as answer]
Thanks for your feedback! 
There was a bug which incorrectly turned off the action "Apply completions to editor". We have prepared a fix, which should be released with the next update.

As long as you have the setting "Preferences > Languages & Framework > GitHub Copilot > Show IDE completions side-by-side" enabled, the fix will allow you to use a custom keyboard shortcut to apply GitHub Copilot's suggestions. As with every action, it's possible to add a 2nd shortcut, e.g. TAB if no popup is shown and something else to apply Copilot's suggestion when a popup is shown.

It's tricky to interfere with the IdeaVim plugin to override handling of ESC, so we'd prefer avoid doing that.

Another option would be to override the IDE's tab handling. But it's likely that some users would prefer the IDE's completion while others prefer GitHub Copilot's suggestions. If you think that the fix doesn't work for your usage patterns, please let us know.
Thanks!

**Reply by kjyv** [reply]
Would be happy to test but there are currently no nightlies being built it seems.
But from what you wrote, I guess it would be an ok solution to allow setting an alternative shortcut that would work while Rider's popup is open and that would work while IdeaVim is in insert mode (I'm not sure what "Show IDE completions side-by-side" is otherwise doing, so can't comment if it is ok for me to be required to enable it).
There is certainly no need to override the handling of ESC as that would be wrong also in terms of the linked IdeaVim ticket, it should in all cases leave insert mode.
My suggestion for what would be good from my initial posting was this:
"So I guess another way that should be usable is that disabling Tab for popup completion should allow Tab to complete Copilot and/or a custom shortcut like Shift-Tab should complete Copilot's suggestion also while the popup is open (and also close it additionally if possible)."
So the second part is what you've fixed and is fine if it also closes Rider's popup while accepting the Copilot suggestion, I will test that once available.
Personally I would prefer being able to use tab for Copilot's suggestion even when the popup is open (not by default as described but if I disable the default shortcut that is set in Rider) as I never accept popup suggestions with Tab, as that seems unnatural to me in terms of usual UI behaviour (where Tab would go to the next UI widget etc.).
